csrc/quantization/fp8_semi_structured/cusparseLt.cpp

```diff
@@ -0,0 +1,600 @@
+      const std::optional<torch::ScalarType> out_dtype_opt) {          \
+    TORCH_CHECK(false, "cusparseLt is not found");                     \
+  }                                                                    \
+  torch::Tensor cslt_mm_semi_structured2(                              \
+      const torch::Tensor& compressed_A, const torch::Tensor& dense_B, \
+      const c10::optional<torch::Tensor>& scale_opt,                   \
+      const c10::optional<torch::Tensor>& bias_opt,                    \
+      const std::optional<torch::ScalarType> out_dtype_opt) {          \
+    TORCH_CHECK(false, "cusparseLt is not found");                     \
+  }                                                                    \
+  void cslt_clear_cache() { TORCH_CHECK(false, "cusparseLt is not found"); }
+
+#if defined(VLLM_CUSPARSELT_ENABLED)
+
+  #include <cusparseLt.h>
+
+  #if defined(CUSPARSELT_VERSION) && CUSPARSELT_VERSION >= 600
+
+    #define CUDASPARSE_CHECK(EXPR)                                 \
+      do {                                                         \
+        cusparseStatus_t __err = EXPR;                             \
+        TORCH_CHECK(__err == CUSPARSE_STATUS_SUCCESS,              \
+                    "CUDA error: ", cusparseGetErrorString(__err), \
+                    " when calling `" #EXPR "`");                  \
+      } while (0)
+
+namespace vllm {
+namespace cusparseLt {
+
+struct cusparseLtEntry {
+  cusparseLtMatDescriptor_t* sparse_input_descriptor_p;
+  cusparseLtMatDescriptor_t* dense_input_descriptor_p;
+  cusparseLtMatDescriptor_t* res_descriptor_p;
+  cusparseLtMatDescriptor_t* C_descriptor_p;
+
+  cusparseLtMatmulDescriptor_t* matmul_p;
+  cusparseLtMatmulPlan_t* plan_p;
+  cusparseLtMatmulAlgSelection_t* alg_sel_p;
+
+  void* workspace_ptr;
+
+  ~cusparseLtEntry() {
+    TORCH_CUDASPARSE_CHECK(
+        cusparseLtMatDescriptorDestroy(sparse_input_descriptor_p));
+    TORCH_CUDASPARSE_CHECK(
+        cusparseLtMatDescriptorDestroy(dense_input_descriptor_p));
+    TORCH_CUDASPARSE_CHECK(cusparseLtMatDescriptorDestroy(C_descriptor_p));
+    TORCH_CUDASPARSE_CHECK(cusparseLtMatDescriptorDestroy(res_descriptor_p));
+    TORCH_CUDASPARSE_CHECK(cusparseLtMatmulPlanDestroy(plan_p));
+
+    // Destructor is called after the cuda cleanup so double free is done here.
+    // AT_CUDA_CHECK(cudaFree(workspace_ptr));
+    delete sparse_input_descriptor_p;
+    delete dense_input_descriptor_p;
+    delete res_descriptor_p;
+    delete C_descriptor_p;
+    delete plan_p;
+    delete alg_sel_p;
+    delete matmul_p;
+  }
+};
+
+cusparseLtHandle_t handle;
+bool handle_initialized = false;
+
+// m, k, n, input_type, output_type, bias enabled, scale enabled, is B
+// contiguous
+using cacheID = std::tuple<int64_t, int64_t, int64_t, at::ScalarType,
+                           at::ScalarType, int, int, int>;
+
+std::map<cacheID, cusparseLtEntry> cusparseLt_cache;
```

Comment:
Ideally the cache entries could be turned into pytorch custom objects but those are not currently supported by the inductor.

Other than using a global map, we could return an opaque pointer to python and embed it in an object with a custom `__del__` method and manage these from python.